### PR TITLE
image load

### PR DESCRIPTION
**Pull request by orhanobut**
- Imageloader is removed, instead imagehandler is introduced.
- Flicker issue is fixed
- Parser is reachable from Wasp now
- LogLevel is reachable from Wasp now.
- Refactoring

### CODE REVIEW

File: app/src/main/java/com/orhanobut/waspsample/ImageHandler.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * @author Orhan Obut
+ */
+public class ImageHandler {
+
+    static RequestQueue requestQueue;
+    static ImageLoader imageLoader;
+
+
```
Comment (AlexBalo) [inline app/src/main/java/com/orhanobut/waspsample/ImageHandler.java]: Too much space :)

File: wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                        if (error != null) {
+                            builder.setNetworkTime(error.getNetworkTimeMs());
+                            errorMessage = error.getMessage();
+
+                            if (error.networkResponse != null) {
+                                NetworkResponse response = error.networkResponse;
+                                String body;
+                                try {
+                                    body = new String(
+                                            error.networkResponse.data, HttpHeaderParser.parseCharset(response.headers)
```
Comment (AlexBalo) [inline wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java]: what about two lines?

File: app/src/main/java/com/orhanobut/waspsample/ImageHandler.java (hunk 1 above)
Comment (orhanobut) [inline app/src/main/java/com/orhanobut/waspsample/ImageHandler.java]: I'll remove them :)

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        imageView.setTag(KEY_TAG, url);
+
+        int width = imageView.getWidth();
+        int height = imageView.getHeight();
+
+        boolean wrapWidth = false;
+        boolean wrapHeight = false;
+        if (imageView.getLayoutParams() != null) {
+            wrapWidth = imageView.getLayoutParams().width == ViewGroup.LayoutParams.WRAP_CONTENT;
+            wrapHeight = imageView.getLayoutParams().height == ViewGroup.LayoutParams.WRAP_CONTENT;
```
Comment (AlexBalo) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: Can we get the LayoutParams just one time and use the reference?

**Comment by AlexBalo** [comment]
Finish reviewing

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java (hunk 3 above)
Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: I agree

File: wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java (hunk 2 above)
Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java]: :+1:

File: app/src/main/java/com/orhanobut/waspsample/ImageHandler.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        if (requestQueue == null) {
+            requestQueue = Volley.newRequestQueue(context);
+            //  imageLoader = new ImageLoader(requestQueue, new BitmapWaspCache());
+        }
+
+        ((NetworkImageView) imageView).setImageUrl(url, imageLoader);
+
+    }
+
+    public static void wasp(Context context, ImageView imageView, String url) {
```
Comment (kardeslik) [inline app/src/main/java/com/orhanobut/waspsample/ImageHandler.java]: Unused parameter "context".

Comment (orhanobut) [inline app/src/main/java/com/orhanobut/waspsample/ImageHandler.java]: This is in app sample side, I think that's why I ignored it.

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            Logger.d("CACHE IMAGE : " + url);
+            return;
+        }
+
+        // make a new request
+        imageNetworkHandler.requestImage(waspImage, maxWidth, maxHeight, new CallBack<Container>() {
+            @Override
+            public void onSuccess(final Container container) {
+                ImageView imageView1 = container.imageView;
+                Bitmap bitmap1 = container.bitmap;
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: naming, no need to append "1".

Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: will fix it

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+    public interface ImageNetworkHandler {
+
+        void requestImage(WaspImage waspImage, int maxWidth, int maxHeight, CallBack<Container> callBack);
+
+        void cancelRequest(String tag);
+
+    }
+
+    public static class Container {
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: Can't we make this and above interfaces package-private?

Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: We can make them package-private.

File: wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * @author Orhan Obut
+ */
+public class VolleyImageNetworkHandler implements WaspImageHandler.ImageNetworkHandler {
+
+    private final RequestQueue requestQueue;
+
+    public VolleyImageNetworkHandler(Context context) {
+        this.requestQueue = Volley.newRequestQueue(context);
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java]: What do you think about using httpStack given to the Wasp.Builder here also? Currently at the time of creating an object from this class we do not have access to the given httpStack but I think we can initialise it when building wasp, can't we?

Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/VolleyImageNetworkHandler.java]: Yes, a very good point. I'll pass the stack here as well.

File: wasp/src/main/java/com/orhanobut/wasp/Wasp.java [hunk 8]
```diff
@@ -51,21 +67,36 @@ private Wasp(Builder builder) {
...
-        private static ImageHandler newImageHandler() {
+        private static ImageHandler getImageHandler() {
             if (context == null) {
                 throw new NullPointerException("Wasp.Builder should be instantiated first");
             }
-            return new VolleyImageHandler(context);
+            if (imageHandler == null) {
+                imageCache = new BitmapWaspCache();
+                imageNetworkHandler = new VolleyImageNetworkHandler(context);
+                imageHandler = new WaspImageHandler(imageCache, imageNetworkHandler);
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/Wasp.java]: We only need imageHandler here, so we can remove other fields and use like this:
<pre>
imageHandler = new WaspImageHandler(new BitmapWaspCache(), new VolleyImageNetworkHandler(context));
</pre>

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            imageView.setImageBitmap(bitmap);
+            Logger.d("CACHE IMAGE : " + url);
+            return;
+        }
+
+        // make a new request
+        imageNetworkHandler.requestImage(waspImage, maxWidth, maxHeight, new CallBack<Container>() {
+            @Override
+            public void onSuccess(final Container container) {
+                ImageView imageView1 = container.imageView;
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: We can move this line to somewhere below for better readability.

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+            // return;
+        }
+
+        // Calculate the max image width / height to use while ignoring WRAP_CONTENT dimens.
+        int maxWidth = wrapWidth ? 0 : width;
+        int maxHeight = wrapHeight ? 0 : height;
+
+        // if there is any old request. cancel it
+        String tag = (String) imageView.getTag(KEY_TAG);
+        if (tag != null) {
```
Comment (kardeslik) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: At line 53, a tag is set to ImageView which makes this if always true. Therefor, this check should be moved to the beginning of method.

By the way, previous request may be with the same url. So, I think we should also check if tag is not equal with current url.

**Comment by kardeslik** [comment]
Done with the review.

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java (hunk 10 above)
Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: This is wrong indeed, check control was at top before setting. By adding width control made this confusion. I'll fix this.

File: wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java (hunk 9 above)
Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/WaspImageHandler.java]: I agree

File: wasp/src/main/java/com/orhanobut/wasp/Wasp.java (hunk 8 above)
Comment (orhanobut) [inline wasp/src/main/java/com/orhanobut/wasp/Wasp.java]: I agree

**Comment by orhanobut** [comment]
pushed the review changes

**Pull request merged**